org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/ClassFileCache.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2018 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.java.resource;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import org.eclipse.emf.common.notify.Adapter;
+import org.eclipse.emf.common.notify.Notifier;
+import org.eclipse.emf.common.notify.impl.AdapterImpl;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.jdt.internal.compiler.env.IBinaryType;
+import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.xbase.lib.Procedures.Procedure2;
+
+/**
+ * A cache for parsed class or java source file content. It is bound to the lifecycle of a resource set.
+ * 
+ * @author Christian Dietrich - Initial contribution and API
+ * @since 2.15
+ */
+public class ClassFileCache {
+	public static class ClassFileCacheAdapter extends AdapterImpl {
+		private ClassFileCache element;
+
+		public ClassFileCacheAdapter(ClassFileCache element) {
+			this.element = element;
+		}
+
+		public ClassFileCache get() {
+			return this.element;
+		}
+
+		@Override
+		public boolean isAdapterForType(Object object) {
+			return object == ClassFileCache.class;
+		}
+	}
+
+	private static final Object NULL = new Object();
+
+	private final Map<QualifiedName, Object> cache = new ConcurrentHashMap<QualifiedName, Object>();
```

Comment:
A few unnecessary type arguments. The diamond op could help to simplify the code.